### PR TITLE
Refactored Piecewise simplification and added simplification methods

### LINKED ISSUE #17283: Eq in condition of Piecewise might be redundant
```python
>>> eq=Piecewise((0, x <= 1), (1, Eq(x, -oo)), (2, True))
>>> simplify(eq)
Piecewise((0, x <= 1), (1, Eq(x, -oo)), (2, True))
```
The result should be `Piecewise((0, x <= 1), (2, True))`: if a condition is an equality sets a previous condition to True then it is redundant. This easy case should probably be removed at instantiation.

(identified [here](https://github.com/sympy/sympy/pull/17146#discussion_r307968690))

### LINKED ISSUE #21481: simplified Piecewise is missing conditions
I have been trying to write a model for the size of symbolic powers.
I define the Piecewise expression as follows for power `b**e` with
the expressions of the Piecewise giving a value that is representative
of the region in which the size of the power is found (e.g. 1/2 if the
value of the power is between 0 and 1):

```python
var('b e')
A=Piecewise((1, Eq(b, 1) | Eq(e, 0) | (Eq(b, -1) & Eq(Mod(e, 2), 0))),
(0, Eq(b, 0) & (e > 0)), (-1, Eq(b, -1) & Eq(Mod(e, 2), 1)),
(Piecewise((2, ((b > 1) & (e > 0)) | ((b > 0) & (b < 1) & (e < 0)) |
((e >= 2) & (b < -1) & Eq(Mod(e, 2), 0)) | ((e <= -2) & (b > -1) & (b
< 0) & Eq(Mod(e, 2), 0))), (S.Half, ((b > 1) & (e < 0)) | ((b > 0) & (e >
0) & (b < 1)) | ((e <= -2) & (b < -1) & Eq(Mod(e, 2), 0)) | ((e >= 2)
& (b > -1) & (b < 0) & Eq(Mod(e, 2), 0))), (-S.Half, Eq(Mod(e, 2), 1) &
(((e <= -1) & (b < -1)) | ((e >= 1) & (b > -1) & (b < 0)))), (-2, ((e
>= 1) & (b < -1) & Eq(Mod(e, 2), 1)) | ((e <= -1) & (b > -1) & (b < 0)
& Eq(Mod(e, 2), 1)))), Eq(im(b), 0) & Eq(im(e), 0)))
```
As a test of folding and simplification I create the following alternative forms for A
```python
B=piecewise_fold(A)
sa=A.simplify()
sb=B.simplify()
```
I then test the forms over a range of `b` and `e` values:
```python
v = Tuple(-2, -1, -0.5, 0, 0.5, 1, 2)
for i in v:
 for j in v:
  r = {b:i,e:j}
  ok=[k.xreplace(r) for k in (A,B,sa,sb)]
  if len(set(ok))!=1:print('ab %s %s'%(r,ok))
```
Although `A` and `B` agree, the simplified forms of each do not and I get these results
```python
s {b: -2, e: -1} [-1/2, -1/2, -1/2, nan]
s {b: -2, e: 1} [-2, -2, -2, nan]
s {b: -1/2, e: 1} [-1/2, -1/2, -1/2, nan]
```
The `nan` indicates that the conditions for the values of `b` and `e` that were passed are not found in the Piecewise expression so a default `nan` is returned. I confirmed that the value returned by `sa` agrees with the value from `A`. I have not yet tracked down the source of the discrepancy from `sb` in which are missing the cases for -2 and -1/2:
```python
Piecewise(
    (1, Eq(b, 1) | Eq(e, 0) | (Eq(b, -1) & Eq(Mod(e, 2), 0))),
    (0, Eq(b, 0) & (e > 0)),
    (-1, Eq(b, -1) & Eq(Mod(e, 2), 1)),
    (2, Eq(im(b), 0) & Eq(im(e), 0) & ((b > 0) | (b > 1)) & ((b > 0) | (e > 0)) & ((b > 1) | (b < 1)) & ((b > 1) | (e < 0)) & ((e > 0) | (b < 1)) & ((e > 0) | (e < 0))),
    (1/2, Eq(im(b), 0) & Eq(im(e), 0) & ((b > 0) | (b > 1)) & ((b > 0) | (e < 0)) & ((b > 1) | (e > 0)) & ((b > 1) | (b < 1)) & ((e > 0) | (e < 0)) & ((b < 1) | (e <0))))
```

### LINKED ISSUE #22066: conflictive expressions in cases statement as integral solution
The following expression
```python
import sympy as sy

x   = sy.symbols("x")
L   = sy.symbols("L", nonzero = True)
k, l = sy.symbols("k, l",   integer=True, nonnegative = True )

integral = sy.Integral( sy.cos( (x*k*sy.pi) / L) * sy.cos( (x*l*sy.pi) / L), (x, -L, L) )
eq = sy.Eq( integral , sy.simplify(integral.doit()) )
eq
```
results in

image

Based on the assumptions given for the symbols the conditions stated in the cases statement are not as expected.
The solution I expected is 
 
image

There are two things that are bothering me

1. cases statement 0 and 1 are conflictive
2. k = -l is always a false statement as k and l where defined as non negative integers

Is there something I'm missing or is this a bug?

### PR DESCRIPTION
**Pull request by oscargus**
#### References to other Issues or PRs

Fixes  #8453
Closes #15929
Closes #17443
Fixes #17283
Closes #21481 (fixed elsewhere)
Fixes #22066

#### Brief description of what is fixed or changed

Refactored the simplification of Piecewise and added more possible simplification steps.

Builds on #15929 and #17443.

One may have to fine tune the steps, but this seems like a reasonable order.

#### Release Notes

* functions
     * Better simplification of `Piecewise`

### CODE REVIEW

**Comment by robinechuca** [comment]
Hello, this looks like the pull request 21933 that I proposed a while ago. I'm glad to hear that the idea has not been abandoned.

**Comment by oscargus** [comment]
@robinechuca This is independent of that (unfortunately) and some old stuff I did a few years back. There is a bit of an issue here which should be merged first as the other will have to be modified quite a bit... But I believe that your additions definitely are worthwhile and should be able to fit in this chain of simplifications somewhere.

**Comment by NotWearingPants** [comment]
Probably also fixes #8453 ?

**Comment by NotWearingPants** [comment]
A few more issues to check (if they're not already handled by this PR then tell me and I'll open issues):

---

```py
Piecewise((0, x>3))
```
simplifies to `0` but it could be `NaN` when `x>=3`, so it should stay as a `Piecewise`.

---

```py
Piecewise(
    ( Piecewise( (1,Eq(x,3)), (2,True) ),    Ne(x,3) ),
    (3, True),
)
```
should realize getting 1 is impossible and simplify to
```py
Piecewise(
    (2, Ne(x,3)),
    (3, True),
)
```

---

```py
Piecewise((x, Ne(x, 1)), (1, True))
```
should simplify to `x`, similar to #8458.

---

btw is the PR still viable or is it too old to fix?

**Comment by oscargus** [comment]
> ```python
> Piecewise((0, x>3))
> ```

This is indeed a bug. I do not think that this PR fixes it (unless coincidental).
> ```python
> Piecewise(
>     ( Piecewise( (1,Eq(x,3)), (2,True) ),    Ne(x,3) ),
>     (3, True),
> )
> ```

`piecewise_fold` will get the expected result. One may for sure consider doing `piecewise_fold` as part of the simplification (I almost thought that it was done).

> ```python
> Piecewise((x, Ne(x, 1)), (1, True))
> ```

There is a at least code in this PR that should simplify it.

> btw is the PR still viable or is it too old to fix?

There is probably a bit of work to sort it out, but I think it would be worthwhile. The problem is if it ends up like this and not getting reviews and then just hangs around. But as 1.11 is about to be released, I guess that one should be able to rebase and merge after that so that there is time to find any issues before the next release.

**Comment by oscargus** [comment]
I rebased it and fixed the conflicts (not pushed yet). It seems like the last case is indeed fixed here. I'll add a test for that as well.

And you are right about #8453. Can't understand how I missed that...

**Comment by oscargus** [comment]
@oscarbenjamin @asmeurer I've rebased this (again) and would like to have it seriously considered for merging. 

It is, naturally, heuristic, and it may be that one must run it more than once etc, but I believe that it is a significant step in terms of simplification capabilities.

**Comment by oscarbenjamin** [comment]
I haven't checked through this in detail but if I understand correctly this is adding a lot of potentially expensive automatic simplification code to `Piecewise.eval`. In general I would rather remove automatic simplification rather than add more of it. I would naturally expect a lot of this to belong in `_eval_simplify` rather than `eval`. The previous `eval` method seems more appropriately restrained in its effort to apply some basic simplification but without working too hard for it.

**Comment by oscargus** [comment]
> lot of potentially expensive automatic simplification code to `Piecewise.eval`

No. It just refactors that part out of `eval` as it is reused in `piecewise_simplify`.

**Comment by oscarbenjamin** [comment]
> > lot of potentially expensive automatic simplification code to `Piecewise.eval`
> 
> No. It just refactors that part out of `eval` as it is reused in `piecewise_simplify`.

Oh sorry, I see what you mean. So `eval` only calls `_piecewise_collapse_arguments` which is equivalent to the existing `eval` code and the other simplification functions are only used in `piecewise_simplify/_eval_simplify`.

File: sympy/logic/boolalg.py [hunk 1]
```diff
@@ -2769,6 +2769,13 @@ def simplify_logic(expr, form=None, deep=True, force=False):
         made. By setting ``force`` to ``True``, this limit is removed. Be
         aware that this can lead to very long simplification times.
 
+    dontcare : Boolean expression
+        Optimize expression under the assumption that this expression is false.
+        This is useful in e.g. Piecewise conditions, where later confitions
+        are subject to the previous conditions being false. Hence, if a previous
+        condition is And(A, B), the simplification of expr can be made with
+        don't cares for And(A, B).
```
Comment (oscarbenjamin) [inline sympy/logic/boolalg.py:2777]: typo -> confitions.

Would it make more sense to negate this and call the argument something like `assumptions`? I think maybe it's easier to understand assuming something to be True rather than assuming something to be False.

Comment (oscarbenjamin) [reply sympy/logic/boolalg.py:2777]: Or perhaps a better name for the argument would be `assume_false`.

Comment (oscargus) [reply sympy/logic/boolalg.py]: I changed the doc-string to better explain what is actually going on.

**Comment by oscargus** [comment]
OK, there are still some bugs here...

```
simplify(Piecewise((x, And(Ne(x, 1), Ne(x, 2))), (1, Eq(x, 1)), (2, Eq(x, 2))))
```
gives 
``` 
⎧x  for x = 1 ∨ x ≠ 2
⎨                    
⎩2      for x = 2    
```

I think that maybe the Ne only can handle the single case at the moment (spent some time thinking about how to deal with this, but cannot really come up with anything efficient, especially considering `And` and `Or`).

File: sympy/logic/boolalg.py (hunk 1 above)
Comment (oscarbenjamin) [reply sympy/logic/boolalg.py]: Okay, that makes more sense.

**Comment by oscarbenjamin** [comment]
This PR helps the example from #21360 by eliminating one case:

Master:
```python
In [1]: a = Symbol('a', real = True)

In [2]: m = (a*y + x**2)*(x**(a-1))

In [3]:  integrate(m.expand(), x).diff(x).simplify()
Out[3]: 
⎧ a - 1 ⎛       2⎞                    
⎪x     ⋅⎝a⋅y + x ⎠  for a ≠ -2 ∧ a ≠ 0
⎪                                     
⎪      a                              
⎪   a⋅x ⋅y + 1                        
⎪   ──────────          for a ≠ 0     
⎪       x                             
⎪                                     
⎨         a + 2                       
⎪  a⋅y + x                            
⎪  ────────────         for a ≠ -2    
⎪       x                             
⎪                                     
⎪     a⋅y + 1                         
⎪     ───────           otherwise     
⎪        x                            
⎩           
```
This PR:
```python
In [3]:  integrate(m.expand(), x).diff(x).simplify()
Out[3]: 
⎧ a - 1 ⎛       2⎞                    
⎪x     ⋅⎝a⋅y + x ⎠  for a ≠ -2 ∧ a ≠ 0
⎪                                     
⎪      a                              
⎨   a⋅x ⋅y + 1                        
⎪   ──────────          for a ≠ 0     
⎪       x                             
⎪                                     
⎩        x              otherwise   
```
It's hard to tell what has caused that change (mixing a refactor with nontrivial changes in the same diff makes the changes hard to follow). It looks as if the third case has been simplified with `a = 0` implying that the condition is True and also that the expression can simplify. What has not simplified is:

- The condition in the second case which could be `a = -2`.
- Then `a = -2` could be used to substitute into the second expression.
- Substituting `a = 0` or `a = -2` into the first expression gives something equivalent to the second and third expressions:
```python
In [4]:  integrate(m.expand(), x).diff(x).simplify().args[0][0].subs(a, -2)
Out[4]: 
 2      
x  - 2⋅y
────────
    3   
   x    

In [5]:  integrate(m.expand(), x).diff(x).simplify().args[1][0].subs(a, -2)
Out[5]: 
    2⋅y
1 - ───
      2
     x 
───────
   x   

In [6]:  integrate(m.expand(), x).diff(x).simplify().args[1][0].subs(a, -2).together()
Out[6]: 
 2      
x  - 2⋅y
────────
    3   
   x    

In [7]:  integrate(m.expand(), x).diff(x).simplify().args[0][0].subs(a, 0)
Out[7]: x
```
If those were handled then the Piecewise could collapse completely.

**Comment by oscarbenjamin** [comment]
This URL gives the (current) diff for this PR minus the initial commit which is the refactor:
https://github.com/sympy/sympy/pull/22753/files/64cedc463177a9a9e218bc0e321208ee6581bc1a..5b67e1298dfad74f30ae5289f05c1bc5ffd2dd23

**Comment by oscargus** [comment]
Well, all the actual simplification takes parts in the different functions, so factoring out one method shouldn't really be a big deal.

**Comment by oscargus** [comment]
> * The condition in the second case which could be `a = -2`.

The thing is that the complexity is not higher for `a != 0`, so `simplify` chooses to not change this. It can of course be hardcoded that a single `Eq` is selected over a single `Ne` if possible.

**Comment by oscargus** [comment]
With a small modification I get:

image

after the first simplification and

image

after the second and finally
image

after the third.

This all comes down to which order one should run the simplifications in.

There will probably always be cases to be found, but I'd go in the direction of improvements. (Heck, if I knew how to make a perfect Piecewise simplification I would work with it. This is probably better than Mathematica for some cases.)

**Comment by oscarbenjamin** [comment]
> This all comes down to which order one should run the simplifications in.

I wonder if it would be simpler to start by applying `piecewise_exclusive` and then manage the different cases more or less independently rather than having each simplification step need to consider the ordering of the cases:
```python
In [8]: piecewise_exclusive(integrate(m.expand(), x).diff(x).simplify())
Out[8]: 
⎧ a - 1 ⎛       2⎞                    
⎪x     ⋅⎝a⋅y + x ⎠  for a ≠ -2 ∧ a ≠ 0
⎪                                     
⎪      a                              
⎪   a⋅x ⋅y + 1                        
⎪   ──────────          for a = -2    
⎪       x                             
⎪                                     
⎨         a + 2                       
⎪  a⋅y + x                            
⎪  ────────────         for a = 0     
⎪       x                             
⎪                                     
⎪     a⋅y + 1                         
⎪     ───────       for a = -2 ∧ a = 0
⎪        x                            
⎩   
```
From here is should be straight-forward to eliminate the final case and substitute values of `a` in the others. In this form we more or less know that Eq and Ne come in pairs so that an Eq in one case must clash with an Ne somewhere else.

**Comment by oscargus** [comment]
Yes, that probably works well for that type of simpler conditions with Ne and Eq. For other cases it is probably a waste of time. But the problem is that it is really hard to know which the correct things to do are for the general case. But one can for sure add some optimizations for that case.

**Comment by oscarbenjamin** [comment]
After the last commit it seems that `Eq` is preferred over `Ne` but the final case is not filtered out any more:
```python
In [9]: simplify(integrate(m.expand(), x).diff(x))
Out[9]: 
⎧ a - 1 ⎛       2⎞                    
⎪x     ⋅⎝a⋅y + x ⎠  for a ≠ -2 ∧ a ≠ 0
⎪                                     
⎪         2⋅y                         
⎪     1 - ───                         
⎪           2                         
⎪          x                          
⎨     ───────           for a = -2    
⎪        x                            
⎪                                     
⎪        x              for a = 0     
⎪                                     
⎪     a⋅y + 1                         
⎪     ───────           otherwise     
⎩        x  
```

**Comment by oscarbenjamin** [comment]
I tried to find a paper about simplifying Piecewise. There don't seem to be many but I did find these:
https://arxiv.org/abs/cs/0702010
https://www.sciencedirect.com/science/article/pii/S0377042714005123

It doesn't look like either of them really handles the case where more than one variable appears in the conditions.

**Comment by oscargus** [comment]
Yeah, you need to run it multiple times at the moment.

Maybe the best thing is to split this in multiple PRs. At the moment I am just adding features rather than making it converge...

So start with refactoring and then add new simplification functions as we go along.

**Comment by oscarbenjamin** [comment]
> Maybe the best thing is to split this in multiple PRs.

It would certainly be easier to review that way.

> At the moment I am just adding features rather than making it converge...

Maybe some of these operations just need to be done in a while loop. I find it hard to follow precisely what all of them do so I'm not sure which. Since each of them runs a pass over the cases maybe it's better to frame it as a single outer pass that builds a stack and tries all methods to merge each case with the previous case until complete. Something like this maybe:

```python
def _piecewise_simplify_args(args):
    first, remaining = args[0], args[1:]
    argstack = [first]
    for case in remaining:
        newcases = [case]
        # Keep popping off the stack while we can merge cases
        while len(newcases) == 1:
            newcases = _merge_cases([argstack.pop(), newcases[0]])
        argstack.extend(newcases)
    return argstack

def _merge_cases(cases):
    while True:
        oldcases = cases
        for func in _merge_funcs:
            cases = func(cases)
            if cases != old_cases:
                break
        else:
            return cases
        
_merge_funcs = [
    _merge_cases_and_or,
    _merge_cases_ne,
    ...
]
```
Then I guess `_piecewise_simplify_args` could do any book-keeping to keep track of acccumulated conditions etc and pass those through to `_merge_cases`.
